app/services/bulk_unsubscribe_service.rb

```diff
@@ -0,0 +1,89 @@
+module BulkUnsubscribeService
+  def self.call(content_ids_and_replacements)
+    affected_subscriber_list_ids = content_ids_and_replacements
+                                     .keys
+                                     .flat_map do |content_id|
+      SubscriberList.find_by_links_value(content_id).pluck(:id)
+    end
+
+    subscriptions_to_end = Subscription
+                             .where(subscriber_list_id: affected_subscriber_list_ids)
+                             .includes(:subscriber)
+                             .active
+
+    subscriptions_to_end
+      .group_by(&:subscriber)
+      .sort_by { |_subscriber, subscriptions| subscriptions.length }
+      .each_with_index do |(subscriber, subscriptions), index|
+
+      subscription_details = subscriptions.map do |subscription|
+        relevant_content_id = subscription
+                                .subscriber_list
+                                .links
+                                .values
+                                .flatten
+                                .find do |links_content_id|
+          content_ids_and_replacements.key? links_content_id
+        end
+
+        [
+          subscription.subscriber_list.title,
+          content_ids_and_replacements.fetch(relevant_content_id)
+        ]
+      end
+
+      email = send_email_to_subscriber(subscriber, subscription_details, index)
+
+      subscriptions.each do |subscription|
+        subscription.update!(
```

Comment:
You could maybe use UnsubscribeService.subscription! here.
It doesn't yet handle the case of unsubscribing with :unpublished but  that is easily added. And you could update Subscription#end as well to reflect this new use case.

Reply:
Using Subscription#end sounds good, but I'm not too convinced about using UnsubscribeService.subscription! here, mostly because passing through the email might be a bit awkward, as it already has some special behaviour about spam tracking.

Reply:
That is a private method. the public one doesn't do spam tracking.

I'm just a bit worried that we end up with multiple paths to unsubscribe a subscription, which makes changing the behaviour of unsubscribing a subscription more difficult

Reply:
> I'm just a bit worried that we end up with multiple paths to unsubscribe a subscription, which makes changing the behaviour of unsubscribing a subscription more difficult

Sure, I've pushed a few commits now which get the BulkUnsubscribeService to use the same code path through the UnsubscribeService. How does that look?